WIP Eeglab matlab hdf5

smgutstein opened the pull request:
Changes to allow mne-python to read files saved by Matlab versions 7.3 or later, which use hdf5.  This handle the first of the 2 problems that PR #4056 is attempting to fix.

Closes #4828

agramfort commented:
can you take one of our test files, convert it to hdf5 format so we can check the data read are the same? thx

larsoner commented:
@agramfort IIRC this was already done a while ago, we have `test_raw_hdf5.set` in the mne-testing-data repo

mmagnuski commented:
Travis also shows pep and docstyle errors:
```
mne/io/eeglab/eeglab.py:46:5: E722 do not use bare except'
mne/io/eeglab/eeglab.py:416:5: E722 do not use bare except'
mne/io/eeglab/eeglab.py:830:5: E722 do not use bare except'
```

```
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D202: No blank lines allowed after function docstring (found 1)
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D205: 1 blank line required between summary line and description (found 0)
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D208: Docstring is over-indented
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D209: Multi-line docstring closing quotes should be on a separate line
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D210: No whitespaces allowed surrounding docstring text
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D400: First line should end with a period (not 'y')
./mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D401: First line should be in imperative mood ('Take', not 'Takes')
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D202: No blank lines allowed after function docstring (found 1)
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D205: 1 blank line required between summary line and description (found 0)
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D208: Docstring is over-indented
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D209: Multi-line docstring closing quotes should be on a separate line
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D210: No whitespaces allowed surrounding docstring text
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D400: First line should end with a period (not 'y')
./build/lib/mne/io/eeglab/eeglab.py:309 in private function `_hdf_data_2_strs`:
        D401: First line should be in imperative mood ('Take', not 'Takes')
```

smgutstein commented:
@larsoner - I did upload  `test_raw_hdf5.set` file, but it will have to be replaced. It was the file I was using in the original PR, which allowed both the reading of hdf5 files, and the use of non-eeg channels. 

Since this PR is just for hdf5 files, I need to replace test_raw_hdf5.set` with the file I've edited to remove the non-eeg channels. Before uploading, I want to edit out some events, since I currently get the following warning:

`test5.py:11: RuntimeWarning: 4 events will be dropped because they occur on the same time sample as another event. `mne.io.Raw` objects store events on an event channel, which cannot represent two events on the same sample. You can extract the original event structure using `mne.io.eeglab.read_events_eeglab`. Then, you can e.g. subset the extracted events for constructing epochs.`

Possibly, this isn't a big deal, but it bugs me.

@mmagnuski - I'll fix those errors. I'm a little puzzled, since everything passed when I ran `make flake` on my machine - perhaps I need to update flake?

@agramfort - This is a little embarrassing, but so far, I've been employing a slightly Neanderthal style of testing:

```
curr_file = 'test_hdf5_raw_min_eeg_only.set'
montage = mne.channels.read_montage('biosemi256')
raw = mne.io.read_raw_eeglab(curr_file) #need to fix reading of curr_file

raw.plot(block=True)
raw.plot_sensors(kind='3d', ch_type='eeg', ch_groups='position', block=True)

```

and as long as things didn't crash and I saw 'resonable' plots, I thought everything was probably good. Is this reasonable, or should I shadow/repeat the tests done in `test_eeglab.py` for `test_raw.set` ?

agramfort commented:
yes you need to update test_eeglab.py and run the same test for each file
(non HDF5 and with HDF5)

let us know if you need more help

smgutstein commented:
Thanks. test_eeglab.py has 4 .set files for testing - test_raw.set, test_raw_onefile.set, test_epochs.set & test_epochs_onefile.set. Am I correct in believing that I only need to shadow the tests applied to test_raw.set ?

agramfort commented:
yes only  test_raw.set
and  test_epochs.set if you can read also such files.

zhang470 commented:
I just tested the most recent MNE io functions (development version as of Feb. 4, 2018) and found that the EEGLAB .set files saved with MATLAB v7.3 still showed the same error as reported earlier. The problem is linked with the io functionality in the mio.py file in scipy which does not seem to be able to handle the compressed hdf5 format at the current stage. But one simple solution for now is to use the following codes in Matlab to rewrite the EEGLAB .set files from v7.3 to v6. Assuming the MATLAB working directory is set for 'xxx.set' written in v7.3, you can do the following: 
% Rewrite .set file in version 6 for
EEG = pop_loadset ('xxx.set');
pop_saveset(EEG, 'filename', 'xxx.set', 'version', '6');

After the rewriting, the io functions for EEGLAB in the MNE package worked just fine. 

You can avoid using these two lines since EEGLAB allows the user to specify the .set output file in either v6 or v7.3. You can modify the customizable 'eeg_options.m' or simply use the graphical interface to set up "Memory and other options" under the 'File' menu so that all .set files are saved in matlab v6 format for maximal compatibility. See the image below. 
![image](https://user-images.githubusercontent.com/5948937/35835789-55018e02-0aa2-11e8-93d0-8ea880219d18.png)

agramfort commented:
@zhang470 did you test this PR or current master?

@smgutstein you need to rebase this branch. thanks

smgutstein commented:
@agramfort - OK. Also, in terms of testing since I don't really know everything that would need to be tested for for scratch, what I think I should do is either:

1. Duplicate every test done on 'test_raw.set', but this seems wasteful
2. Just write a bunch of tests to ensure that each field from my input file is read as I expect it to be. I'm just worried that this test could be a little self-fulfilling and brittle. 


Other than that, I really don't know an appropriate way of testing. Do you have any ideas/preferences?

larsoner commented:
Can you accomplish (1) by indenting the existing code to use a test loop over the two filenames? If the outputs are equivalent at the EEGLAB/MATLAB end, then this should be doable and require little effort (indent + add loop, replace fixed filename with loop filename) at the test end.

Maybe you don't need to modify *every* test that uses the EEGLAB file, just one (or two) that cover core functionality.

agramfort commented:
@smgutstein can you share with me the hdf5 set files you need to use? I will add them to the mne-testing-data repo.

smgutstein commented:
Sure thing. I'll do it at the same time I have the unit tests going.

On Wed, Feb 7, 2018 at 3:06 AM, Alexandre Gramfort <notifications@github.com
> wrote:

> @smgutstein <https://github.com/smgutstein> can you share with me the
> hdf5 set files you need to use? I will add them to the mne-testing-data
> repo.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-363688239>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJIaDVImm5koDLN6hJh-ME0QTnV1uks5tSVmFgaJpZM4Rd1Y_>
> .
>

smgutstein commented:
@larsoner, @agramfort  - I believe the unit tests I need to run are as follows:

In test_eeglab.py, I need to repeat the tests done for raw_fname ('test_raw.set')  in the sections commented by "#main tests, and test_missing_id", "#test finding events in continuous data" and "#test that using uint16_codec does not break stuff". Is this correct? 

Also, since the test data I used was hand edited (i.e. reduced) from some actual lab data, it may not be appropriate for these tests (I honestly don't know one way or the other). What I believe I should do is include in the unit test a conversion of the 'test_raw.set' file to temp file using the hdf format, and to read converted file. This should also save the overhead of saving another data file just for an infrequently used unit test.

Does this seem reasonable?

agramfort commented:
the easy root for me is to:
- read in matlab all our .set files we use for testing. Save them as v73
format in Hdf5.
- add all the files to mne-testing-data
- run the same tests on these new files with a simple for loop. "for fname
in [fname_v4, fname_v73]: ..."

agramfort commented:
@smgutstein you will need to rebase.

tell us when you added the tests.

smgutstein commented:
Note: I still need to add the tests to test_eeglab. I just want to see if this code passes the auto checks and get whatever code-style comments people feel inclined to make

smgutstein commented:
Note: I still need to add the tests to test_eeglab. I just want to see if this code passes the auto checks and get whatever code-style comments people feel inclined to make

agramfort commented:
I usually work the opposite 

I write tests and for you it's touching 15 lines of code 

Then I make Travis happy and then I ask for reviews

agramfort commented:
one easy option for you is something like this

rename test_io_set to _test_io_set

def _test_io_set(use_hdf5):
      if use_hdf5:
            fname = "test_set_v73.set"
      else:
            fname = "test_set.mat"

      # rest stayas the same

def test_io_set():
      _test_io_set(use_hdf5=True)
      _test_io_set(use_hdf5=False)

ok?

smgutstein commented:
Am making modifications to test_eeglab along @agramfort suggestion, but have question.  test_eeglab use io.savemat to save datawhen I read in data stored in hdf5 format, I put it into a Bunch object instead of io.matlab.mio5_params.mat_struct, so io.savemat no longer works. Possible soln's (that I see) are:

1. Introduce a _save_eeg_data method to eeglab.py, in addition to _get_eeg_data method and use this for all saving of data.

2. Instead of putting hdf5 data into Bunch object, put it into a io.matlab.mio5_params.mat_struct object.

Option 1 would be fairly quick & easy for me and would encapsulate things more, but option 2 might make for more self-consistency, since the class in which eeg data is stored won't be dependent upon the file format it was stored in.

Any preferences?

agramfort commented:
you should not test / support saving back to matlab .set file. You just
want to read hdf5 files.

smgutstein commented:
But test_eeglab has io.savemat statements in lines 117, 135, 148, 169, 199,
238 & 263. If I don't want it to crash, I need to save data within
test_eeglab.py. The purpose for most of these saves seem to be to read the
data back immediately, but I'm not entirely certain.

On Mon, Mar 12, 2018 at 11:52 AM, Alexandre Gramfort <
notifications@github.com> wrote:

> you should not test / support saving back to matlab .set file. You just
> want to read hdf5 files.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-372359461>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJEqVWyuDwmI6oJYGt0TuXVsy7FQNks5tdpm4gaJpZM4Rd1Y_>
> .
>

agramfort commented:
hum we have these savemat to be able to test corner cases without adding
more files in the testing repo.
maybe split the test_eeglab function in 2 and only test hdf5 with the
splitted function that does not use savemat

smgutstein commented:
OK - Or taking your earlier suggestion, I'll just use the 'use_hdf5' flag
to branch around the savemats

On Mon, Mar 12, 2018 at 12:24 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> hum we have these savemat to be able to test corner cases without adding
> more files in the testing repo.
> maybe split the test_eeglab function in 2 and only test hdf5 with the
> splitted function that does not use savemat
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-372371031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJKxndEBCqkxH8x6T416_UxK2crpTks5tdqEzgaJpZM4Rd1Y_>
> .
>

agramfort commented:
ok

agramfort commented:
@smgutstein please see why travis and appveyor are not happy

smgutstein commented:
I'm not entirely certain what's causing the problem. The main complaint is an inability to find the test files:
```

E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_raw_h5.set'
E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_raw_h5.set.mat'
E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_epochs_h5.set'
E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_epochs_h5.set.mat'
E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_raw_onefile_h5.set'
E           FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\appveyor\\mne_data\\MNE-testing-data\\EEGLAB\\test_raw_onefile_h5.set.mat'
E           FileNotFoundError: [Errno 2] No such file or directory: '/home/travis/mne_data/MNE-testing-data/EEGLAB/test_raw_h5.set'
E           FileNotFoundError: [Errno 2] No such file or directory: '/home/travis/mne_data/MNE-testing-data/EEGLAB/test_raw_h5.set.mat'
E           FileNotFoundError: [Errno 2] No such file or directory: '/home/travis/mne_data/MNE-testing-data/EEGLAB/test_epochs_h5.set'
E           FileNotFoundError: [Errno 2] No such file or directory: '/home/travis/mne_data/MNE-testing-data/EEGLAB/test_epochs_h5.set.mat'

```
etc.

I tried recloning the mne-testing-data git and noticed that it created the directory `mne-testing-data` instead of `MNE-testing-data`. To get the tests working on my system again, I needed to change the directory name. I don't know how this change was introduced, nor what to do about it. I can see that in line 301 of mne/datasets/utils.py `MNE-testing-data` seems to be expected, and all the other folder/directory names have the prefix `MNE`, not `mne`.  I don't know whether the correct change is to change all the folder/directory names in util.py so that they use `mne`, or to change the way the MNE-testing-data git clones itself so that it installs itself into `MNE-testing-data`

agramfort commented:
does it work locally for you when you download a clean version of the
testing data?

larsoner commented:
> I tried recloning the mne-testing-data git

This is different from what is done in `utils.py`. Typically you don't clone but rather `mne.datasets.testing.data_path(force_update=True, verbose=True)`. This will name the directory properly.

larsoner commented:
... at least for the step of testing code you don't `git clone`. You have two options:

1. Clone `mne-testing-data` for the purposes of changing `mne-testing-data` when you need to, but keep a separate copy that you got from `mne.datasets.testing.data_path()`.
2. Clone and rename locally the `mne-testing-data` directory `MNE-testing-data`.

larsoner commented:
... it sounds like you've done the second one. To change "the way it clones itself", we would either need to change the repo name (not a good idea) or add to the README that you can do something like this:
```
git clone https://github.com/mne-tools/mne-testing-data MNE-testing-data
```
But in any case, that shouldn't be why this PR is failing.

larsoner commented:
This PR does not update the `mne/datasets/utils.py` version tag or hash. Can you try step (5) of the instructions to update?

https://github.com/mne-tools/mne-testing-data/blob/master/README.md

smgutstein commented:
When I download a clean version of the testing data, using `mne.datasets.testing.data_path(force_update=True, verbose=True).`, everything works (at least until it notices I'm using numpy 1.14, but that's something else). I used hash = `a1568d9d238887aa2942f138307cf89a` and release = `0.45`. This also installed a directory with the expected name - `MNE-testing-data`

larsoner commented:
Great, bump the version tag and update the hash in `mne/datasets/utils.py` in this PR and the CIs should get the update, too

smgutstein commented:
I'm having trouble understanding the errors I'm seeing. Mainly I don't quite know why they don't show up when I run things locally.

In appeyvor, the one error seems to come from failing the following test:

mne\io\eeglab\tests\test_eeglab.py:87: in _test_io_set
assert_equal(len(w), 4)

I assume what happened was something went wrong when I was converting an array of ascii values into a string (or list of strings), but I don't see an error when I run locally.

The errors in travis seem to mainly be not being able to import h5py, and complaining about raising a NotImplementedError, which I thought my try/except blocks were catching

larsoner commented:
Checking warning counts is unreliable, though we have some older testing code that does it. It's better to check to ensure the content you want is actually there, e.g.:
```
assert any('my warning' in str(ww.message) for ww in w)
```

larsoner commented:
For the h5py error, if the test requires h5py, add the `requires_h5py` decorator from `mne.utils` to the `test_` function.

smgutstein commented:
@larsoner - Well, I think I've found the problem, but I'm not entirely certain what to do about it. 

I determine whether or not a file is hdf5 or not using a try/except block. When io.loadmat/io.whomat throws a "Not Implemented" error, it doesn't close the file it opened. This leads to all the extra warnings I see (i..e. in line ~89: assert_equal(len(w), 4) of test_eeglab.py  Error: 14 != 4). I do not know why these warnings occur in Python 3.5 and not Python 2.7, but that's another issue. 

I don't really know how best to close these files. When I fall into the except block to read an hdf file, I no longer have the file descriptor that was left open by io.loadmat. I could try to use subprocess and the lsof command, but that's only good for Linux. I tried using psutil to find open files, but when I closed the file opened by io.loadmat before reopening it with h5py.File, I got an error when I tried to close the file (re)opened with h5py. (I hope this is clear)

Now, my only thought is to go through the list of warnings and filter out these warnings: 

{message : ResourceWarning("unclosed file <_io.BufferedReader name='/home/me/mne_data/MNE-testing-data/EEGLAB/test_raw_h5.set'>",), category : 'ResourceWarning' ...

Is there a better approach to try? It seems there should be some way to close these files safely.

agramfort commented:
looking at :
https://github.com/scipy/scipy/blob/master/scipy/io/matlab/mio.py#L19

it seems you can pass a file_like to loadmat so maybe you can open the file
before loadmat
so you have a fid to close?

larsoner commented:
You'll need to use the `requires_h5py` decorator for tests that use h5py.

Then you'll need to fix `flake` errors:

https://travis-ci.org/mne-tools/mne-python/jobs/360288365

But the main functionality tests pass, which is good!

larsoner commented on mne/io/base.py:
```diff
@@ -657,7 +657,10 @@ def annotations(self):  # noqa: D401
+        try:
+            return tuple(self._filenames)
+        except AttributeError:
+            return tuple([None])
```
Don't change `base` change `eeglab.py` to set this to `[None]`, usually done in `super(RawEEGLAB, self).__init__(...)`, see e.g.:

https://github.com/mne-tools/mne-python/blob/master/mne/io/edf/edf.py#L169

Also this is a bit weird to set it to `None`, it should point to the underlying `.set` / `.mat` file so it's displayed properly e.g. in `raw.plot`.

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -34,14 +35,50 @@ def _check_fname(fname):
-                     squeeze_me=True)
+    try:
+        # Try to read old style Matlab file
+        fhandle = open(fname, 'rb')
```
```
with open(fname, 'rb') as fid:
    mat = ...
```
then no need for close

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -34,14 +35,50 @@ def _check_fname(fname):
+    except NotImplementedError:
+        # Try to read new style Matlab file
+        fhandle.close()
+        import h5py
```
```
try:
    import h5py
except ImportError:
    raise RuntimeError('Reading v7+ MATLAB format .set files requires h5py, which could not be imported')
```

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -34,14 +35,50 @@ def _check_fname(fname):
+        # Try to read new style Matlab file
+        fhandle.close()
+        import h5py
+        f = h5py.File(fname)
```
```
with h5py.File(fname) as f:
    mat = list(f.keys())
```
then no need for close

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +327,247 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+    sz = len(derefs[list(derefs.keys())[0]])
+    bch_list = [Bunch(**{key: derefs[key][x] for key in derefs})
+                for x in range(sz)]
+    return bch_list
```
This code looks like it was all taken from somewhere else, basically to convert modern `.mat` files to the same format legacy `.mat` files would yield when used with `scipy.io.loadmat`, correct? If so, put all of these functions in `mne/externals/something.py`

larsoner replied:
... this comment is the one I reference above

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +327,247 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+            if c3:
+                eeg[curr_field] = ''.join([chr(y) for y in temp])
+
+    return eeg
```
This function, which does the triaging and a tiny bit of custom conversion should stay here.

Make sense?

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -354,6 +662,18 @@ def __init__(self, input_fname, montage, eog=(), event_id=None,
+            else:
+                hdf5_transpose = True
+        else:
+            hdf5_transpose = False
```
It looks now like conversion is scattered -- is that correct? This for example could be moved to `_get_eeg_data` so that all conversion / unification takes place in one location.

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -369,10 +689,18 @@ def __init__(self, input_fname, montage, eog=(), event_id=None,
+            # Seem to have transpose with matlab hdf storage
+            if hdf5_transpose:
+                temp = eeg.data.transpose()
+                eeg.data = temp
```
same thing -- if data needs to be transposed, can you do it in the above function so that this one can stay "clean"?

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -654,8 +1000,22 @@ def read_events_eeglab(eeg, event_id=None, event_id_func='strip_to_integer',
+            f = h5py.File(eeg)
+            eeg_dict = hdf_2_dict(f, f['EEG'], parent=None)
+            f.close()
+            eeg = _bunchify(eeg_dict)
```
This looks redundant with code that is above. Can you refactor into a private function that is used both places?

larsoner commented on mne/io/eeglab/tests/test_eeglab.py:
```diff
@@ -35,8 +36,28 @@
+    global raw_fname
+    global raw_fname_onefile
+    global epochs_fname
+    global epochs_fname_onefile
```
Avoid using `global` where possible. Call these filenames `epochs_fname_h5` etc. at the top with the others, then triage here which one to load.

smgutstein replied:
I incorporated the structural suggestions. Tests pass on my machine, but we'll see what happens when I push.

I undid the change to base.py, because I can't recreate the error that necessitated it (which is weird, but so was finding something that needed to be fixed in base.py). The weird part of the error was that I couldn't figure out how a rewEEGLAB object managed to avoid having a filename attribute assigned in the first place.

I removed the globals, but am still not certain I should have. I was using the globals to overshadow some module wide variables within a local method, without unduly effecting the other methods in the modules, in this case, since it's a unit test module, but maybe I'm wrong.

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +330,272 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+
+                bunch_data[ctr].__dict__[curr_field] = deref
+
+    return bunch_data
```
See previous comment about putting this is `externals` and adding a link to whatever code this was taken from (assuming it was adapted from somewhere -- judging by the docstring I assume it was)

larsoner replied:
Let's not put it in externals then. Let's make a new file `mne/io/eeglab/_h5.py`. Move `_get_hdf_eeg_data` there (and all necessary related functions, such as the bunchify business), and if everything is refactored properly, the only import in this file that should be necessary is `from ._h5 import _get_hdf_eeg_data`. This will force the h5-related refactoring into one place (`_h5.py`) and keep the (unified) EEGLAB->MNE translation here. Make sense?

If not -- or if you are getting burned out on the refactoring -- then I can take a stab at it.

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +330,272 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+                           'be imported')
+
+    logger.info("Attempting to read Matlab style hdf file")
+    f = h5py.File(input_fname)
```
Missed previous comments about using `with open` context manager rather than `open / close`

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +330,272 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+    except ImportError:
+        raise RuntimeError('Reading v7+ MATLAB format .set',
+                           'requires h5py, which could not',
+                           'be imported')
```
This check several places, it should be possible to do it only one place I think

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -354,6 +690,13 @@ def __init__(self, input_fname, montage, eog=(), event_id=None,
+            # as array of ascii values of characters
+            ascii_check = _check_for_ascii_filename(eeg, input_fname)
+            if ascii_check[0]:
+                eeg.data = ascii_check[1]
```
Can't you put thsi is `_get_eeg_data`? Then it doesn't need to be replicated here and in the Epochs reader.

larsoner commented on mne/io/eeglab/tests/test_eeglab.py:
```diff
@@ -35,8 +36,29 @@
+        raw_fname = op.join(base_dir, 'test_raw.set')
+        raw_fname_onefile = op.join(base_dir, 'test_raw_onefile.set')
+        epochs_fname = op.join(base_dir, 'test_epochs.set')
+        epochs_fname_onefile = op.join(base_dir, 'test_epochs_onefile.set')
```
What I suggest is to put all of filenames (old format and new format) above, around line 34. It's probably worth renaming the old variables to e.g. `raw_fname_old` or `raw_fname_legacy` for the old/legacy MATLAB format. Then here you do:
```
if use_hdf:
    raw_fname = raw_fname_h5
    ...
else:
    raw_fname = raw_fname_old
    ...
```

smgutstein commented:
The code you want to put in mne_externals is basically mine. I based
something on a gist that recursively converted nested dicts to nested named
tuples, which were eventually replaced with MNE-Python's bunches. That
said, it may still make sense to put the conversion code in some other mne
sub-directory. I don't know.

I, also, was bugged by the repetition of the try/except clauses for
importing h5py, but since I don't really know the possible orders in which
the methods can be called, I left them all in. Is there a better
alternative?

I'm not certain about putting both the old & new format files as module
scope variables. There are two other test-suites in the module that use the
existing module scope variables, but don't test with hdf5 files.
test_degenerate explicitly doesn't test with hdf5 files and
test_eeglab_annotations doesn't seem to need testing with the new files,
since read_annotations_eeglab uses _get_eeg_data, but maybe I'm wrong about
this.



On Sat, Mar 31, 2018 at 1:00 PM, Eric Larson <notifications@github.com>
wrote:

> *@larsoner* commented on this pull request.
>
> Looks cleaner, but it looks like it can be more unified. See also the
> unaddressed comment about mne/externals
> ------------------------------
>
> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436560>:
>
> > +        c4 = (isinstance(bd[0], h5py.h5r.Reference))
> +        if (c1 and c2 and c3 and c4):
> +            for ctr in range(len(bunch_data)):
> +                bd = bunch_data[ctr].__dict__[curr_field]
> +                try:
> +                    # Ensure bunch_data[ctr].__dict__[curr_field] is iterable
> +                    # before attempting to iterate over it
> +                    iter(bd)
> +                except TypeError:
> +                    deref = [orig[bd].value.flatten()]
> +                else:
> +                    deref = [orig[x].value.flatten() for x in bd]
> +
> +                bunch_data[ctr].__dict__[curr_field] = deref
> +
> +    return bunch_data
>
> See previous comment about putting this is externals and adding a link to
> whatever code this was taken from (assuming it was adapted from somewhere
> -- judging by the docstring I assume it was)
> ------------------------------
>
> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436573>:
>
> > +                bunch_data[ctr].__dict__[curr_field] = deref
> +
> +    return bunch_data
> +
> +
> +def _get_hdf_eeg_data(input_fname):
> +
> +    try:
> +        import h5py
> +    except ImportError:
> +        raise RuntimeError('Reading v7+ MATLAB format .set',
> +                           'requires h5py, which could not',
> +                           'be imported')
> +
> +    logger.info("Attempting to read Matlab style hdf file")
> +    f = h5py.File(input_fname)
>
> Missed previous comments about using with open context manager rather
> than open / close
> ------------------------------
>
> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436587>:
>
> > +            str_data = _bunch_data_2_strs(bunch_data, curr_field)
> +            for ctr, curr_str in enumerate(str_data):
> +                bunch_data[ctr].__dict__[curr_field] = curr_str
> +    return bunch_data
> +
> +
> +def _bunch_derefs(orig, bunch_data, deref_fields):
> +    """ Dereferences h5py.h5r.Reference objects. Ensures that each
> +        field of bunch object with dereferenced objects stores them
> +        in a list, even if that list has only 1 element. """
> +    try:
> +        import h5py
> +    except ImportError:
> +        raise RuntimeError('Reading v7+ MATLAB format .set',
> +                           'requires h5py, which could not',
> +                           'be imported')
>
> This check several places, it should be possible to do it only one place I
> think
> ------------------------------
>
> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436596>:
>
> > +                  isinstance(h5_values[x][0], h5py.Reference)
> +                  else h5_values[x]
> +                  for x in h5_values}
> +
> +        for ct in in_hlGroup:
> +            msg = indent + "Converting " + tuple_name + '_' + ct
> +            logger.debug(msg)
> +
> +    except IOError:
> +        derefs = {ct: [None] for ct in in_hlGroup}
> +        warn("Couldn't read", tuple_name, ". Assuming empty")
> +
> +    sz = len(derefs[list(derefs.keys())[0]])
> +    bch_list = [Bunch(**{key: derefs[key][x] for key in derefs})
> +                for x in range(sz)]
> +    return bch_list
>
> ... this comment is the one I reference above
> ------------------------------
>
> In mne/io/eeglab/eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436612>:
>
> > @@ -354,6 +690,13 @@ def __init__(self, input_fname, montage, eog=(), event_id=None,
>          self._create_event_ch(events, n_samples=eeg.pnts)
>
>          # read the data
> +        if isinstance(eeg, Bunch):
> +            # if hdf5 data was read in, then file name might be encoded
> +            # as array of ascii values of characters
> +            ascii_check = _check_for_ascii_filename(eeg, input_fname)
> +            if ascii_check[0]:
> +                eeg.data = ascii_check[1]
>
> Can't you put thsi is _get_eeg_data? Then it doesn't need to be
> replicated here and in the Epochs reader.
> ------------------------------
>
> In mne/io/eeglab/tests/test_eeglab.py
> <https://github.com/mne-tools/mne-python/pull/4874#discussion_r178436667>:
>
> > +
> +@testing.requires_testing_data
> +@requires_h5py
> +def _test_io_set(use_hdf=True):
> +    """Test importing EEGLAB .set files."""
> +
> +    if use_hdf:
> +        raw_fname = op.join(base_dir, 'test_raw_h5.set')
> +        raw_fname_onefile = op.join(base_dir, 'test_raw_onefile_h5.set')
> +        epochs_fname = op.join(base_dir, 'test_epochs_h5.set')
> +        epochs_fname_onefile = op.join(base_dir, 'test_epochs_onefile_h5.set')
> +    else:
> +        raw_fname = op.join(base_dir, 'test_raw.set')
> +        raw_fname_onefile = op.join(base_dir, 'test_raw_onefile.set')
> +        epochs_fname = op.join(base_dir, 'test_epochs.set')
> +        epochs_fname_onefile = op.join(base_dir, 'test_epochs_onefile.set')
>
> What I suggest is to put all of filenames (old format and new format)
> above, around line 34. It's probably worth renaming the old variables to
> e.g. raw_fname_old or raw_fname_legacy for the old/legacy MATLAB format.
> Then here you do:
>
> if use_hdf:
>     raw_fname = raw_fname_h5
>     ...
> else:
>     raw_fname = raw_fname_old
>     ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#pullrequestreview-108497153>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJAuCNjRyOnAuits6uLu4NsiPKmF3ks5tj7ZGgaJpZM4Rd1Y_>
> .
>

larsoner commented:
> I'm not certain about putting both the old & new format files as module
> scope variables. There are two other test-suites in the module that use the
> existing module scope variables, but don't test with hdf5 files.
> test_degenerate explicitly doesn't test with hdf5 files and
> test_eeglab_annotations doesn't seem to need testing with the new files,
> since read_annotations_eeglab uses _get_eeg_data, but maybe I'm wrong about
> this.

It is fine to rename the filename variables used by these files to reflect the new names. It is cleaner than doing tricks with variable scope or (worse) global vars.

agramfort commented:
thx @smgutstein 

can you just look at coverage report https://codecov.io/gh/mne-tools/mne-python/compare/5f53bd96450e138bf98605d0d90d8018651926d2...c33121f84de95eb2cbbc72dee94a252dc3442679/src/mne/io/edf/edf.py

and see why it went down a bit?

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -53,8 +93,23 @@ def _check_mat_struct(fname):
+        if isinstance(ll, np.ndarray):
+            return list(ll) if ll.size > 0 else np.nan
+        else:
+            return ll
```
Can this all be replaced with:
```
ll = np.array(ll, float).ravel()
ll = np.nan if ll.size == 0 else ll
return ll
```
?

larsoner commented on mne/io/eeglab/eeglab.py:
```diff
@@ -260,6 +330,284 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
+    except ImportError:
+        raise RuntimeError('Reading v7+ MATLAB format .set',
+                           'requires h5py, which could not',
+                           'be imported')
```
I think that this is the only place this check should be done, i.e. when you try to first load the data in HDF5 format. All the others should be removed / removable.

smgutstein commented:
@larsoner  - Let me take one more stab at it this weekend. I think I'd like to take the last few steps on this. I appreciate all the support you guys have given me in doing this, and just hope my part-time learning project for myself has saved you more time/effort than it generated.

@agramfort - I tried looking at the coverage report, but didn't really understand what I was looking at. It's somewaht puzzling to me that edt.py was effected at all, since I don't see how it interacts with use of eeglab. If you're willing, I'd like to punt this back to you or at least ask for some pointers as to what it is I'm looking for.(sorry).

larsoner commented:
> It's somewaht puzzling to me that edt.py was effected at all

Yes sometimes the coverage reports can be a bit iffy. Concentrate on `eeglab.py` where your changes are, see if there are lines that are never used:

https://codecov.io/gh/mne-tools/mne-python/pull/4874/src/mne/io/eeglab/eeglab.py?before=mne/io/eeglab/eeglab.py

Based on looking at it, it looks like your HDF5-related changes are not tested, is that possible? Maybe `codecov` is messing up again.

agramfort commented:
my bad I meant to look indeed at eeglab.py ...

can you test the coverage locally?

with somethign like this:

nosetests mne/io/eeglab/ --with-coverage --cover-package=mne.io.eeglab

smgutstein commented:
@agramfort  @larsoner - OK, that makes sense.  Since I'm going to be refactoring most (almost all) of my changes out of eeglab.py, I'll do that first, then start staring at the coverage.

smgutstein commented:
OK - I guess it's done. Hope I saved more work than I caused.

larsoner commented:
@smgutstein did you forget to push a commit? Last commit I see was 7 days ago, there are unaddressed comments (e.g., the `hdf5`-specific stuff wasn't moved), etc.

> Hope I saved more work than I caused.

We're making good progress, thanks for sticking with it!

smgutstein commented:
@larsoner - No, I was getting set to address the comments this morning, but saw the PR was marked "closed" and added to the 0.16 milestone, so I thought you might have decided to finish it up. If I make changes and then push here, am I sending changes I make to the right place? Am I duplicating work that is being done on the 0.16 milestone?

larsoner commented:
The PR is still open, yes please continue if you have time

smgutstein commented:
Testing splitting out of h5-related code from eeglab.py Will look at coverage once this is done.

agramfort commented on mne/io/eeglab/eeglab.py:
```diff
@@ -72,32 +104,47 @@ def _get_info(eeg, montage, eog=()):
+            info = create_info(ch_names, eeg.srate, ch_types='eeg')
+            if n_channels_with_pos > 0:
+                selection = np.arange(n_channels_with_pos)
+                montage = Montage(np.array(pos), pos_ch_names, kind, selection)
```
why is the montage part only for non HDF5 file?

smgutstein replied:
As I recall, my thinking was that the hdf5_flag got thrown when the hdf5 file didn't store chanlocs (i.e. eeg.chanlocs is empty). So the montage part would be best handled by the `if montage is None` section, since I used the default channel names after the TypeError was thrown and hdf_flag set to True. Was that reasonable?

agramfort replied:
please check as I really don't see why this can be necessary. an old .set file or a new hdf5 contain exactly the same information. So besides the reading to get something that looks like the output of io.loadmat I really don't see why the code should be different.

to see if there is an option to share the maintenance load can you see if https://gitlab.com/obob/pymatreader does a good job for eeglab files?

smgutstein replied:
They do contain the same information, it's just an empty eeg.chanlocs was stored as array([0,0], dtype=uint64), which does not have a zero length. So, when you look at the code (from _get_info, lines 72-103)

    if len(eeg.chanlocs) > 0: #This branch is where an empty eeg.chanlocs from hdf5 file wrongly goes
        pos_fields = ['X', 'Y', 'Z']
        if (isinstance(eeg.chanlocs, np.ndarray) and not isinstance(
                eeg.chanlocs[0], io.matlab.mio5_params.mat_struct)):
            has_pos = all(fld in eeg.chanlocs[0].dtype.names
                          for fld in pos_fields)
        else:
            has_pos = all(hasattr(eeg.chanlocs[0], fld)
                          for fld in pos_fields)
        get_pos = has_pos and montage is None
        pos_ch_names, ch_names, pos = list(), list(), list()
        kind = 'user_defined'
        update_ch_names = False
        for chanloc in eeg.chanlocs:
            ch_names.append(chanloc.labels)
            if get_pos:
                loc_x = _to_loc(chanloc.X)
                loc_y = _to_loc(chanloc.Y)
                loc_z = _to_loc(chanloc.Z)
                locs = np.r_[-loc_y, loc_x, loc_z]
                if not np.any(np.isnan(locs)):
                    pos_ch_names.append(chanloc.labels)
                    pos.append(locs)
        n_channels_with_pos = len(pos_ch_names)
        info = create_info(ch_names, eeg.srate, ch_types='eeg')
        if n_channels_with_pos > 0:
            selection = np.arange(n_channels_with_pos)
            montage = Montage(np.array(pos), pos_ch_names, kind, selection)
    elif isinstance(montage, string_types):
        path = op.dirname(montage)
    else:  # if eeg.chanlocs is empty, we still need default chan names
             # This is where empty eeg.chanlocs from hdf5 file should go 
        ch_names = ["EEG %03d" % ii for ii in range(eeg.nbchan)]

I really needed to redirect the code execution to the `else` branch of the if, since it had incorrectly landed in the `if` branch. I used the resulting `TypeError` thrown by the eeg.chanlocs variable, which was mistakenly thought to be not-empty, to execute the code in the `else` block and to create a flag that would let me exit the 'if' branch of the `if-then-else` block.

smgutstein replied:
I'll look at `pymatreader` later this week.

smgutstein replied:
I was thinking last night that, perhaps, the real answer is in _h5.py to check for whether eeg.chanlocs should really be empty and then, if necessary, make it so. This way the logic in _get_info can return to its original form.

agramfort commented:
test time for eeglab code went from 2s to 37s with this PR.

it's a lot. can this be improved?

smgutstein commented:
I'll have to check. That's a bizarrely large increase.

On Mon, Apr 9, 2018 at 3:11 AM, Alexandre Gramfort <notifications@github.com
> wrote:

> test time for eeglab code went from 2s to 37s with this PR.
>
> it's a lot. can this be improved?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-379656451>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJNUPyivFK62R9DYSuAbntDEBhOleks5tmwm3gaJpZM4Rd1Y_>
> .
>

agramfort commented:
I agree with your nightly thoughts :)

larsoner commented:
@smgutstein any chance to look and see if we can put a copy of `pymatreader` as `mne/externals/pymatreader` and use it here?

https://gitlab.com/obob/pymatreader

larsoner commented:
(Then any inconsistencies between the old and new file types can be opened as upstream bugs, fixed there, and our internal version updated)

cbrnr commented:
+1 for using `pymatreader`! I assume this means that any locally installed version of that package takes precedence over the one in `mne/externals`?

agramfort commented:
yes via the mechanism of relative imports

larsoner commented:
For uses *inside* the mne packge, the one in `mne/externals` will always be used, if that's what you're asking.

larsoner commented:
Moving this to 0.17 because we probably want `pymatreader` in first

larsoner commented:
@smgutstein if you rebase now you should be able to use `pymatreader`. Are you still interested in giving it a shot, or have you tired of our shifting requirements? :)

smgutstein commented:
I'm still interested, but a little short on time for the next few weeks. I
have to admit though, I'm feeling a certain Ahab-like obsession with this
......

On Wed, May 16, 2018 at 11:35 AM, Eric Larson <notifications@github.com>
wrote:

> @smgutstein <https://github.com/smgutstein> if you rebase now you should
> be able to use pymatreader. Are you still interested in giving it a shot,
> or have you tired of our shifting requirements? :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-389563881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJMQh4pgDNMOg_BBkQTvTX9eaSv6Yks5tzEczgaJpZM4Rd1Y_>
> .
>

larsoner commented:
@smgutstein still interested or should someone else take over? I think a new PR might be the way to go because the steps should probably be:

1. Reconfigure existing EEGLAB readers to use `pymatreader`
2. Add tests for the H5 files, which should immediately work if `pymatreader` has done its job properly

smgutstein commented:
Hi Eric,

Sorry, but I'm sort of running from pillar to post lately. I still feel an
Ahab like compulsion just to get the damn thing done, but don't really know
that I'll have the time.

Steven

On Wed, Jun 27, 2018 at 12:25 PM, Eric Larson <notifications@github.com>
wrote:

> @smgutstein <https://github.com/smgutstein> still interested or should
> someone else take over? I think a new PR might be the way to go because the
> steps should probably be:
>
>    1. Reconfigure existing EEGLAB readers to use pymatreader
>    2. Add tests for the H5 files, which should immediately work if
>    pymatreader has done its job properly
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4874#issuecomment-400740381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfaJBScB_6IiuhM8VVkiwlou39B4v_qks5uA7HrgaJpZM4Rd1Y_>
> .
>

larsoner commented:
Okay I'll take over, thanks for the efforts @smgutstein !
